src/FSharp.Core/prim-types.fs

```diff
@@ -5206,6 +5217,29 @@ namespace Microsoft.FSharp.Core
                 if value = unsafeDefault<'T> then ""
                 else let x = (# "" value : IFormattable #) in defaultIfNull "" (x.ToString(null, CultureInfo.InvariantCulture))
 
+        [<CompiledName("ToString")>]
+        let inline string (value: 'T) = 
+             defaultString value
+
+             // Only compilers that understand `when 'T : SupportsWhenTEnum` will understand `when 'T : Enum`.
+             when 'T : SupportsWhenTEnum =
+                (
+                    let inline string (value : 'T) =
+                        defaultString value
+
+                        // Special handling for enums whose underlying type is a signed integral type.
+                        // The runtime value may be outside the defined members of the enum, and the negative sign may be overridden.
+                        when 'T : Enum  = let x = (# "" value : 'T #) in x.ToString() // Use 'T to constrain the call to the specific enum type.
+
+                        // For compilers that understand `when 'T : Enum`, we can safely make a constrained call on the integral type itself here.
+                        when 'T : sbyte = let x = (# "" value : sbyte #) in x.ToString(null, CultureInfo.InvariantCulture)
```

Comment:
Is this block fully superceeding the existing block for integral types?
In what constellations is the old block still being hit?


https://github.com/dotnet/fsharp/pull/18546/files#diff-06bb7bec1ea92205ec6f4e69f53cdf3aa53e3431db58ca15a30c7d426ab6f24bR5189-R5192

Reply:
This supersedes the existing block for enums and signed integral types only when compiled by a compiler that understands `when 'T : SupportsTEnum` and `when 'T : Enum`. Older compilers will not give special processing to `when 'T : SupportsTEnum`, so nothing will match that rule (they will treat it as an exact type equivalence test, which they understand but which nothing can ever match), and they will immediately fall back to `defaultString`, which contains the old sequence of static optimizations.

Reply:
Got it.
In that case please adjust the comment over at L 5176.
The code should still remain, but the reasons changed.

You can also add e.g. a 5 year timer in that comment (or current date + that this block of code is needed for the supported lifetime of .NET8 SDKs)